src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala

```diff
@@ -0,0 +1,130 @@
+            val errorMessage = {
+              val baseMessage = "%d failures out of %d attempts saving profile.  Errors: %s"
+                .format(profilePropertyMap.size, errors.size, errors mkString ",")
+              if (missingReports.isEmpty) baseMessage
+              else {
+                val supplementalErrorMessage = "Additionally, %d of these failures did not provide error reports: %s"
+                  .format(missingReports.size, missingReports mkString ",")
+                baseMessage + "\n" + supplementalErrorMessage
+              }
+            }
+            Future(RequestCompleteWithErrorReport(InternalServerError, errorMessage, errorReports))
+        }
+      } recover { case e: Throwable => RequestCompleteWithErrorReport(InternalServerError, e.getMessage) }
+
+      profileResponse pipeTo context.parent
+
+  }
+
+  def updateUserProperties(userInfo: UserInfo, profilePropertyMap: Map[String, String]): Future[List[HttpResponse]] = {
+    val pipeline = authHeaders(requestContext) ~> sendReceive
+    val propertyPosts = profilePropertyMap map {
+      case (key, value) =>
+        pipeline {
+          Post(UserService.remoteSetKeyURL,
+            ThurloeKeyValue(
+              Some(userInfo.getUniqueId),
+              Some(FireCloudKeyValue(Some(key), Some(value)))
+          ))
+        }
+    }
+    Future sequence propertyPosts.toList
+  }
+
+  def checkUserInRawls(requestContext: RequestContext): Future[PerRequestMessage] = {
+    val pipeline = authHeaders(requestContext) ~> sendReceive
+    pipeline { Get(UserService.rawlsGetUserURL) } flatMap { response =>
+        response.status match {
+          case x if x == OK =>
+            Future(RequestComplete(OK))
+          case x if x == NotFound =>
+            registerUserInRawls(requestContext)
+          case _ =>
+            Future(RequestCompleteWithErrorReport(response.status, response.toString))
+        }
+    } recover { case e: Throwable => RequestCompleteWithErrorReport(InternalServerError, e.getMessage) }
+  }
+
+  def registerUserInRawls(requestContext: RequestContext): Future[PerRequestMessage] = {
+    val pipeline = authHeaders(requestContext) ~> sendReceive
+    pipeline { Post(UserService.rawlsRegisterUserURL) } map { response =>
+        response.status match {
+          case x if x.isSuccess || isConflict(response) =>
+            RequestComplete(OK)
+          case _ =>
+            RequestCompleteWithErrorReport(response.status, response.toString)
+        }
+    } recover { case e: Throwable => RequestCompleteWithErrorReport(InternalServerError, e.getMessage) }
+  }
+
+  /**
+    * Rawls can come back with a conflict (500 wrapping a 409) if the user exists either in
+    * Rawls or LDAP when registering. That is not a real error so we can consider that a success.
+    *
+    * @param postResponse The HttpResponse
+    * @return A conflict or not
+    */
+  def isConflict(postResponse: HttpResponse): Boolean = {
```

Comment:
Is there a corner case where the user exists in LDAP but not in rawls, where the isConflict will return false because the response entity doesn't contain "409 Conflict"? Maybe have to look for either "409 Conflict" or "LDAP: error code 68"?

I wish there were a more precise way to check for this case, but given the response from rawls, I'm not seeing it.